docker/entrypoint.sh

```diff
@@ -0,0 +1,43 @@
+#!/bin/sh
+
+# usage: file_env VAR [DEFAULT]
+#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
+# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
+#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
+file_env() {
+    local var="$1"
+    local fileVar="${var}_FILE"
+    local def="${2:-}"
+    if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
+        echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
+        exit 1
+    fi
+    local val="$def"
+    if [ "${!var:-}" ]; then
+        val="${!var}"
+    elif [ "${!fileVar:-}" ]; then
+        val="$(< "${!fileVar}")"
+    fi
+    export "$var"="$val"
+    unset "$fileVar"
+}
+
+file_env 'IOTA_AUTH_USER'
+file_env 'IOTA_AUTH_PASSWORD'
+
+# Reading username and password at runtime
```

Comment:
@Jagatjot  - I forgot this one:

This section is incorrect and needs amendment. You don't need to read Username and password here and it doesn't resolve the previous comments.

1. If you are not running in secure mode, warn
2. If you are running in secure mode and haven't added Username/Password warn

The text about a difficult pattern is incorrect. The Username/Password will be defined for you.

Reply:
I have kept username/password to null in this https://github.com/telefonicaid/iotagent-json/pull/380/commits/151094fb44fab57ec2ba01e6b9c170c31ceb3c42 and printed a warning to set the auth credentials?

Reply:
The text 

> It is recommended to set IOTA Auth credentials(using difficult pattern) 

Still doesn't make sense - The Username/Password will be defined for you.

These lines:

```bash
export IOTA_AUTH_PASSWORD=$IOTA_AUTH_PASSWORD
export IOTA_AUTH_USER=$IOTA_AUTH_USER
```

repeat the work of the  `fileEnv()` function  - delete


This if-then

```bash
if [[ $IOTA_AUTH_USER ==  "" ]] || [[ $IOTA_AUTH_PASSWORD == "" ]]; then
```

needs to be an if-then-else - see https://github.com/telefonicaid/iotagent-json/pull/380#discussion_r254205868

* First check if the Auth is enabled (and if not echo a warning)
* Then for users in secure mode only check if  IOTA_AUTH_USER + IOTA_AUTH_PASSWORD are set (not blank) and echo the other warning.

If I am not running in secure mode I don;t want an irrelevant warning about setting a username I don't need.

Reply:
> @jason-fox from where the username/password be coming in entrypoint.sh file. Will it come from config.js file if we do not read them here ?

They are supplied as  `ENV` variables when instantiating the docker container - see [this example](https://github.com/telefonicaid/iotagent-json/tree/master/docker). Or indeed as **Docker Secrets** - encrypted files with an associated `ENV` variable - in other words the whole idea of this issue - avoiding placing plain text sensitive data into a file (or an `ENV` which can then be inspected.

If you use the **Docker Secrets** method, all you get is the file path to an encrypted file.

If you don't use `ENV` variables then the `config.js` values would be used as default You could of course use your own `config.js` (by attaching it as a `VOLUME`) - but once again you'll end up with the passwords in plain text within a file.

Reply:
Fixed with https://github.com/telefonicaid/iotagent-json/pull/380/commits/efc36a520e965bd17fe0b21d9ac7bb202431504c

Reply:
Fixed with https://github.com/telefonicaid/iotagent-json/pull/380/commits/c6aa3ab7d521f562f849e863dc83272cffdbce37

Reply:
@jason-fox I have updated the changes in https://github.com/telefonicaid/iotagent-json/pull/380/commits/c6aa3ab7d521f562f849e863dc83272cffdbce37 . It first check if auth is enabled or not(if not, its gives a warning to use authentication), if it is enabled but the username and password are not set(it gives the warning to set username and password) and if username/password are set, it gives the warning to use difficult passwords. Thank you!
I